### PR TITLE
feature/gainers losers

### PR DESCRIPTION
**Pull request by devanshu-santiment**
This PR has additions for new gainers and losers page.

The table looks like this: 
screencapture-localhost-3000-social-movers-2019-04-02-16_29_42

Also we have a menu link under the labs dropdown "Social movers" to get to this page

### CODE REVIEW

File: src/ducks/GainersAndLosers/GainersAndLosersTable.js [hunk 1]
```diff
@@ -1,69 +1,63 @@
 import React from 'react'
 import ReactTable from 'react-table'
-import 'react-table/react-table.css'
+import moment from 'moment'
+import { graphql } from 'react-apollo'
 import getColumns from './gainers-and-losers-table-columns'
-import { fetchGainersOrLosers } from './actions'
-import { connect } from 'react-redux'
-import { bindActionCreators } from 'redux'
+import { TOP_SOCIAL_GAINERS_LOSERS_GQL } from './gainersLosersGQL'
```
Comment (Partysun) [inline src/ducks/GainersAndLosers/GainersAndLosersTable.js]: Instead of TOP_SOCIAL_GAINERS_LOSERS_GQL
TOP_SOCIAL_GAINERS_LOSERS_QUERY as Apollo team does. We make it for all new code.

File: src/utils/utils.js [hunk 2]
```diff
@@ -297,6 +297,21 @@ const pickFork = (...forks) => props => {
  */
 const isEthStrictAddress = address => /^0x[0-9a-f]{40}$/i.test(address)
 
+const getTimeRangeByDuration = timeWindow => {
```
Comment (Partysun) [inline src/utils/utils.js]: Add spec @devanshu-santiment

File: src/utils/utils.js [hunk 3]
```diff
@@ -297,6 +297,21 @@ const pickFork = (...forks) => props => {
...
+  const unit = timeWindow.slice(timeWindow.length - 1)
+  const timeAmount = timeWindow.slice(0, timeWindow.length - 1)
```
Comment (DmitriVanGuard) [inline src/utils/utils.js]: `.slice` accepts negative numbers. Please, read [documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/slice#Using_slice()_with_negative_indexes)

**Comment by devanshu-santiment** [comment]
@Partysun @DmitriVanGuard should this page have a link in menu bar? and where?

**Comment by Partysun** [comment]
@devanshu-santiment Add the link in Labs dropdown.

**Comment by devanshu-santiment** [comment]
@Partysun few project don't have images in assets/projects. I propose using a div with fix size as default image that will align the Project names properly cc @Partysun 

image

**Comment by Partysun** [comment]
@devanshu-santiment Ask Andrey, He will give you a dummy icon.

**Comment by Partysun** [comment]
@devanshu-santiment Make this layout 
image

**Comment by Partysun** [comment]
For 24 and 2 use 
image
@devanshu-santiment

**Comment by Partysun** [comment]
@devanshu-santiment Remove a grid of each mini chart.

File: src/components/Navbar/NavbarLabsDropdown.js [hunk 4]
```diff
@@ -8,7 +8,8 @@ const links = [
   { link: '/ethereum-spent', label: 'ETH spent' },
   { link: '/labs/balance', label: 'Historical balance' },
   { link: '/labs/wordcloud', label: 'Word context' },
-  { link: '/dashboards', label: 'Dashboard' }
+  { link: '/dashboards', label: 'Dashboard' },
+  { link: '/social-movers', label: 'Social movers' }
```
Comment (Partysun) [inline src/components/Navbar/NavbarLabsDropdown.js]: /labs/social-movers @devanshu-santiment

File: src/components/ProjectIcon/index.js [hunk 5]
```diff
@@ -1,26 +1,34 @@
...
 
-export const ProjectIcon = ({ name, size, ticker, className }) => {
+export const ProjectIcon = ({
+  name,
+  size,
+  ticker,
+  className,
+  useFallbackIcon
+}) => {
+  const Icon = useFallbackIcon ? DefaultProjectIcon : DefaultIcon
```
Comment (Partysun) [inline src/components/ProjectIcon/index.js]: @devanshu-santiment Why do we need to have useFallbackIcon. Can we understand do we have this icon or not inside this class?

File: src/components/ProjectLabel.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+import ProjectIcon from './ProjectIcon'
+import styles from './ProjectLabel.module.scss'
+
+const ProjectLabel = ({ name, ticker, className }) => (
+  <div className={cx(styles.label, className)}>
+    <ProjectIcon
+      name={name}
+      ticker={ticker}
+      size={20}
+      className='project-icon-default'
```
Comment (Partysun) [inline src/components/ProjectLabel.js]: Why not css module style? @devanshu-santiment

File: src/components/ProjectLabel.js [hunk 7]
```diff
@@ -0,0 +1,27 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import cx from 'classnames'
+import ProjectIcon from './ProjectIcon'
+import styles from './ProjectLabel.module.scss'
+
+const ProjectLabel = ({ name, ticker, className }) => (
```
Comment (Partysun) [inline src/components/ProjectLabel.js]: @devanshu-santiment add story in our storybook for this component.

File: src/ducks/GainersAndLosers/GainersAndLosersTable.js [hunk 8]
```diff
@@ -0,0 +1,67 @@
+import React from 'react'
+import ReactTable from 'react-table'
+import { graphql } from 'react-apollo'
+import getColumns from './gainers-and-losers-table-columns'
+import { TOP_SOCIAL_GAINERS_LOSERS_GQL } from './gainersLosersGQL'
+import { getTimeRangeByDuration } from '../../utils/utils'
+import 'react-table/react-table.css'
+
+const GainersLosersTable = ({
+  isLoading,
```
Comment (Partysun) [inline src/ducks/GainersAndLosers/GainersAndLosersTable.js]: What is an error @devanshu-santiment ?

File: src/utils/utils.js [hunk 9]
```diff
@@ -297,6 +297,32 @@ const pickFork = (...forks) => props => {
...
+      .subtract(timeAmount, unit)
+      .toISOString(),
+    timeWindow
+  }
+}
+
+const mapItemsToKeys = (items, { keyPath, getKeyPath }) => {
+  const obj = {}
+
+  items.forEach(item => {
```
Comment (Partysun) [inline src/utils/utils.js]: use reduce if you want to return one variable obj. @devanshu-santiment 
Also, it fixed mutation of const obj variable. Sure it's JS and it's ok, but better to keep the rule if const than it's const.

File: src/pages/Detailed/DetailedHeader.js [hunk 10]
```diff
@@ -4,7 +4,7 @@ import {
   createSkeletonElement
 } from '@trainline/react-skeletor'
 import { compose } from 'recompose'
-import ProjectIcon from './../../components/ProjectIcon'
+import ProjectIcon from '../../components/ProjectIcon'
```
Comment (Partysun) [inline src/pages/Detailed/DetailedHeader.js]: Why did you remove dot / @devanshu-santiment .
Below paths have this pattern.

File: src/ducks/GainersAndLosers/GainersLosersPage.js [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+const withProjects = graphql(allProjectsGQL, {
+  props: ({ data: { allProjects = [], loading } }) => ({
+    allProjects: !loading
+      ? mapItemsToKeys(allProjects, { keyPath: 'slug' })
+      : undefined,
+    isLoadingProjects: loading
+  })
+})
+
+export default withProjects(GainersAndLosersPage)
```
Comment (Partysun) [inline src/ducks/GainersAndLosers/GainersLosersPage.js]: @devanshu-santiment better name withProject = enhance.
Why? 
Because sometimes you can make a name this as withApollo or smth similar, but we have such HOCs in libs or utils

File: src/ducks/GainersAndLosers/GainersAndLosersTable.js [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+      status,
+      size: 10000,
+      ...getTimeRangeByDuration(timeWindow)
+    }
+  }),
+  props: ({ data: { topSocialGainersLosers = [], loading }, ownProps }) => {
+    let mappedGainersLosers = []
+
+    if (!loading) {
+      const { projects } = topSocialGainersLosers.pop() || {}
```
Comment (Partysun) [inline src/ducks/GainersAndLosers/GainersAndLosersTable.js]: is it only one element in this array? @devanshu-santiment

Comment (devanshu-santiment) [reply src/ducks/GainersAndLosers/GainersAndLosersTable.js]: @Partysun getting an array from `topSocialGainersLosers` query sorted by date incrementally. I suppose we will have to show the data from the most recent records we have.

File: src/components/ProjectLabel.module.scss [hunk 13]
```diff
@@ -0,0 +1,19 @@
+.label {
+  width: 100%;
+  text-align: left;
+  padding: 0 10px;
+  display: flex;
+  align-items: center;
+}
+
+.name {
+  color: var(--black);
```
Comment (DmitriVanGuard) [inline src/components/ProjectLabel.module.scss]: Old css-var colors are deprecated and there was a warning in discord. Please, use the naming of the up-to-date color scheme

File: src/components/ProjectLabel.module.scss [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+  padding-left: 10px;
+  font-size: 14px;
+}
+
+.ticker {
+  padding-left: 4px;
+  color: #7a859e;
```
Comment (DmitriVanGuard) [inline src/components/ProjectLabel.module.scss]: Should be a css-var

File: src/ducks/GainersAndLosers/GainersLosersPage.js [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+
+const timeWindowMapping = {
+  '24h': '2d',
+  '2w': '15d'
+}
+
+class GainersAndLosersPage extends Component {
+  state = {
+    timeWindow: '2d',
+    status: 'GAINER'
```
Comment (DmitriVanGuard) [inline src/ducks/GainersAndLosers/GainersLosersPage.js]: `statusMapping` and `timeWindowMapping` created above, while the state is constructed with literal values.

File: src/ducks/GainersAndLosers/GainersLosersPage.js [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+    const { allProjects } = this.props
+    return (
+      <div className='page'>
+        <h1>Top Gainers And Top Losers</h1>
+        <Panel className={styles.table}>
+          <div className={styles.tabs}>
+            <Tabs
+              options={['Gainers', 'Losers']}
+              defaultSelectedIndex='Gainers'
+              onSelect={selectedTab =>
```
Comment (DmitriVanGuard) [inline src/ducks/GainersAndLosers/GainersLosersPage.js]: **[D-1-1]** Should be a class instance method, e.g. `onTabSelect`, otherwise why to create dynamically allocated closures in the class component

File: src/ducks/GainersAndLosers/GainersLosersPage.js [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+                this.setState({
+                  status: statusMapping[selectedTab]
+                })
+              }
+            />
+            <Tabs
+              options={['24h', '2w']}
+              defaultSelectedIndex='24h'
+              onSelect={selectedTab =>
```
Comment (DmitriVanGuard) [inline src/ducks/GainersAndLosers/GainersLosersPage.js]: Same as **[D-1-1]**

File: src/ducks/GainersAndLosers/GainersAndLosersTable.js (hunk 12 above)
Comment (Partysun) [reply src/ducks/GainersAndLosers/GainersAndLosersTable.js]: ok

**Comment by devanshu-santiment** [comment]
> For 24 and 2 use
> image
> @devanshu-santiment

@Partysun  For this, in existing code we are using [formik](https://jaredpalmer.com/formik/)
Question is we can do it in two ways-
1. Use formik and style that inside santiment/app
2. I can add a variant of Tab component an san-ui

I think 2. is better, what you think?

File: src/components/ProjectLabel.module.scss (hunk 13 above)
Comment (devanshu-santiment) [reply src/components/ProjectLabel.module.scss]: @DmitriVanGurad
black doesn't seems to be mapped with any new var. I've done the changes for other colors

**Comment by Partysun** [comment]
@devanshu-santiment We have this component in san-ui https://storybook.stage.internal.santiment.net/?selectedKind=Selector&selectedStory=Simple&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybook%2Factions%2Factions-panel

Formik component is a wrapper for using this component in Formik form.

File: src/components/ProjectLabel.js (hunk 6 above)
Comment (devanshu-santiment) [reply src/components/ProjectLabel.js]: I realised we don't need that anyway

File: src/components/ProjectIcon/index.js (hunk 5 above)
Comment (devanshu-santiment) [reply src/components/ProjectIcon/index.js]: Had a conversation with Andriy and got to know that we will be using the same default icon for assets table as well. So I replaced that default Icon instead

File: src/components/ProjectLabel.module.scss (hunk 13 above)
Comment (DmitriVanGuard) [reply src/components/ProjectLabel.module.scss]: @devanshu-santiment If the color palette does not have a color with the corresponding name, just check hex color
image

File: src/components/ProjectLabel.js (hunk 6 above)
Comment (DmitriVanGuard) [reply src/components/ProjectLabel.js]: If there is no need for that anyway, why it's not removed?

File: src/ducks/GainersAndLosers/GainersAndLosersTable.js [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+  error
+}) => (
+  <ReactTable
+    loading={isLoading}
+    showPaginationTop={false}
+    showPaginationBottom
+    defaultPageSize={10}
+    getTheadThProps={() => ({ style: { textAlign: 'center' } })}
+    getTrProps={() => ({ style: { textAlign: 'center' } })}
+    getTdProps={() => ({
```
Comment (DmitriVanGuard) [inline src/ducks/GainersAndLosers/GainersAndLosersTable.js]: **[LOC 20, 21, 22, 29]** Unnecessary closures which are created on every rerender. Should be moved out and created as constants

File: src/ducks/GainersAndLosers/GainersLosersGraph.js [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+  <LineChart
+    width={300}
+    height={50}
+    data={socialGainersLosersStatus}
+    margin={{ top: 5, right: 20, left: 20, bottom: 5 }}
+  >
+    <Line
+      type='monotone'
+      dataKey='change'
+      stroke='var(--grey-dark)'
```
Comment (DmitriVanGuard) [inline src/ducks/GainersAndLosers/GainersLosersGraph.js]: There is no `grey-dark` in our color palette. Please, check the color by hex if the name corresponding does not exist

File: src/ducks/GainersAndLosers/GainersLosersPage.js [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+
+  render () {
+    const { timeWindow, status } = this.state
+    const { allProjects, error } = this.props
+    return (
+      <div className={cx('page', styles.wrapper)}>
+        <h1>Top Gainers And Losers</h1>
+        <div className={styles.tabs}>
+          <Selector
+            options={['Gainers', 'Losers']}
```
Comment (Partysun) [inline src/ducks/GainersAndLosers/GainersLosersPage.js]: Object.keys(statusMapping)

File: src/ducks/GainersAndLosers/GainersLosersPage.js [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+            defaultSelected='Gainers'
+            onSelectOption={this.onSelectStatus}
+          />
+          <Selector
+            options={['24h', '2w']}
```
Comment (Partysun) [inline src/ducks/GainersAndLosers/GainersLosersPage.js]: Object.keys(timeWindowMapping)

File: src/ducks/GainersAndLosers/GainersLosersPage.js [hunk 22]
```diff
@@ -0,0 +1,76 @@
+import React, { Component } from 'react'
+import { graphql } from 'react-apollo'
+import cx from 'classnames'
+import { Selector } from '@santiment-network/ui'
+import GainersLosersTable from './GainersAndLosersTable'
+import { allProjectsGQL } from '../../pages/Projects/allProjectsGQL'
+import { mapItemsToKeys } from '../../utils/utils'
+import styles from './GainersAndLosersPage.module.scss'
+
+const statusMapping = {
```
Comment (Partysun) [inline src/ducks/GainersAndLosers/GainersLosersPage.js]: such consts should be STATUSES and TIME_WINDOWS @devanshu-santiment

File: src/ducks/GainersAndLosers/GainersLosersPage.js (hunk 20 above)
Comment (Partysun) [reply src/ducks/GainersAndLosers/GainersLosersPage.js]: @devanshu-santiment

File: src/ducks/GainersAndLosers/GainersLosersPage.js (hunk 21 above)
Comment (Partysun) [reply src/ducks/GainersAndLosers/GainersLosersPage.js]: @devanshu-santiment

File: src/ducks/GainersAndLosers/GainersAndLosersTable.js [hunk 23]
```diff
@@ -0,0 +1,74 @@
...
+        display: 'flex',
+        justifyContent: 'center',
+        alignItems: 'center'
+      }
+    })}
+    getTbodyProps={() => ({ style: { overflow: 'hidden' } })}
+    pageSizeOptions={[5, 10, 20, 25, 50, 100]}
+    sortable={false}
+    loadingText='Loading...'
+    noDataText={error && 'Error fetching data'}
```
Comment (Partysun) [inline src/ducks/GainersAndLosers/GainersAndLosersTable.js]: isError. (we use isLoading, isError and isDisabled) @devanshu-santiment

File: src/ducks/GainersAndLosers/GainersAndLosersTable.js [hunk 24]
```diff
@@ -0,0 +1,74 @@
...
+    let mappedGainersLosers = []
+
+    if (!loading) {
+      const { projects } = topSocialGainersLosers.pop() || {}
+
+      mappedGainersLosers =
+        ownProps.allProjects && projects
+          ? projects.map(projectItem => ({
+            ...projectItem,
+            name: ownProps.allProjects[projectItem.slug].name
```
Comment (Partysun) [inline src/ducks/GainersAndLosers/GainersAndLosersTable.js]: @devanshu-santiment What if allProjects[projectItem.slug] is undefined?
We will have an error - Cannot read property 'name' of undefined 
Better name for this var is project.

File: src/ducks/GainersAndLosers/GainersAndLosersTable.js [hunk 25]
```diff
@@ -0,0 +1,74 @@
...
+          }))
+          : []
+    }
+
+    return {
+      topSocialGainersLosers: mappedGainersLosers,
+      isLoading: loading,
+      error: ownProps.error || error
```
Comment (Partysun) [inline src/ducks/GainersAndLosers/GainersAndLosersTable.js]: isError @devanshu-santiment

**Comment by devanshu-santiment** [comment]
image

@Partysun updated SS
